### ISSUE TITLE
"Error - Chart Metadata (Chart.yaml) missing" when using a Helm chart as a GitRepo artifacts

### ISSUE THREAD

**Issue by angelcervilla** [spinnaker/spinnaker#5589]
### Issue Summary:
We were implementing some pipelines that consumes our own developed Helm charts stored in the corporate Git repository directly, and we were just trying them out by creating a simple pipeline with only 1 bake stage, and the chart is in a subpath, as you can see here:

![pipeline-setup](https://user-images.githubusercontent.com/44018346/76881691-d0b3d780-6879-11ea-9686-0740f673b812.png)

So, when I proceed to run it, the pipeline fails after a couple of seconds with the message

`Bake failed: Error: chart metadata (Chart.yaml) missing`

So, I went to check Clouddriver logs, and I see that it does the job:
```java
2020-03-17 17:07:10.994  INFO 1 --- [         task-1] c.n.s.c.a.g.GitRepoArtifactCredentials   : Cloning git/repo https://<gitserver>/scm/team1/helm-chart-examples into /tmp/e29e4a1b-08f2-422e-a95e-19ca12f42f8d
2020-03-17 17:07:12.104  INFO 1 --- [         task-1] c.n.s.c.a.g.GitRepoArtifactCredentials   : Creating archive for git/repo https://<gitserver>/scm/team1/helm-chart-examples
```
So, Clouddriver seems to do the job well, and I proceed to look further, and the error comes from Rosco:

```java
2020-03-17 17:07:12.622  INFO 1 --- [0.0-8087-exec-6] c.n.s.rosco.jobs.local.JobExecutorLocal  : Starting job: [helm, template, /tmp/rosco-8699768100136721157/ceb61020-0299-4aea-b386-ade99772db77, --name, sc-nginx, --namespace, test, --values, /tmp/rosco-8699768
100136721157/eff56553-65e9-45ba-aaf3-f6a7b10ff308]...
2020-03-17 17:07:12.622  INFO 1 --- [0.0-8087-exec-6] c.n.s.rosco.jobs.local.JobExecutorLocal  : Polling state for ef59d07a-4d9a-4299-9f9b-4269a9b4ccc7 (executionId: null)...
2020-03-17 17:07:12.622  INFO 1 --- [tionScheduler-1] c.n.s.rosco.jobs.local.JobExecutorLocal  : Executing ef59d07a-4d9a-4299-9f9b-4269a9b4ccc7 with tokenized command: [helm, template, /tmp/rosco-8699768100136721157/ceb61020-0299-4aea-b386-ade99772db77, --name, sc-nginx, --
namespace, test, --values, /tmp/rosco-8699768100136721157/eff56553-65e9-45ba-aaf3-f6a7b10ff308] (executionId: 01E3MQD89J7WAXWCS6NRBCVAEA:01E3MQD8AWABWRF984YHK74W4X:1)
2020-03-17 17:07:13.622  INFO 1 --- [0.0-8087-exec-6] c.n.s.rosco.jobs.local.JobExecutorLocal  : Polling state for ef59d07a-4d9a-4299-9f9b-4269a9b4ccc7 (executionId: 01E3MQD89J7WAXWCS6NRBCVAEA:01E3MQD8AWABWRF984YHK74W4X:1)...
2020-03-17 17:07:13.623  INFO 1 --- [0.0-8087-exec-6] c.n.s.rosco.jobs.local.JobExecutorLocal  : State for ef59d07a-4d9a-4299-9f9b-4269a9b4ccc7 changed with exit code 1 (executionId: 01E3MQD89J7WAXWCS6NRBCVAEA:01E3MQD8AWABWRF984YHK74W4X:1).
2020-03-17 17:07:13.623  INFO 1 --- [0.0-8087-exec-6] c.n.s.rosco.jobs.local.JobExecutorLocal  : StdOut for failed job ef59d07a-4d9a-4299-9f9b-4269a9b4ccc7:
2020-03-17 17:07:13.623  INFO 1 --- [0.0-8087-exec-6] c.n.s.rosco.jobs.local.JobExecutorLocal  : StdErr for failed job ef59d07a-4d9a-4299-9f9b-4269a9b4ccc7: Error: chart metadata (Chart.yaml) missing

2020-03-17 17:07:13.630 ERROR 1 --- [0.0-8087-exec-6] c.n.s.k.w.e.GenericExceptionHandlers     : Internal Server Error

java.lang.IllegalStateException: Bake failed: Error: chart metadata (Chart.yaml) missing

        at com.netflix.spinnaker.rosco.manifests.BakeManifestService.doBake(BakeManifestService.java:63) ~[rosco-manifests.jar:na]
        at com.netflix.spinnaker.rosco.manifests.helm.HelmBakeManifestService.bake(HelmBakeManifestService.java:36) ~[rosco-manifests.jar:na]
        at com.netflix.spinnaker.rosco.manifests.helm.HelmBakeManifestService.bake(HelmBakeManifestService.java:12) ~[rosco-manifests.jar:na]
        at com.netflix.spinnaker.rosco.controllers.V2BakeryController.doBake(V2BakeryController.java:44) ~[rosco-web.jar:na]
        at sun.reflect.GeneratedMethodAccessor116.invoke(Unknown Source) ~[na:na]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_232]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_232]
        at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:190) ~[spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138) ~[spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:104) ~[spring-webmvc-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:892) ~[spring-webmvc-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:797) ~[spring-webmvc-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1039) ~[spring-webmvc-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:942) ~[spring-webmvc-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1005) ~[spring-webmvc-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:908) ~[spring-webmvc-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:660) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:882) ~[spring-webmvc-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:741) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) [tomcat-embed-websocket-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.springframework.web.filter.ShallowEtagHeaderFilter.doFilterInternal(ShallowEtagHeaderFilter.java:108) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at javax.servlet.FilterChain$doFilter.call(Unknown Source) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47) [groovy-2.5.7.jar:2.5.7]
        at javax.servlet.FilterChain$doFilter.call(Unknown Source) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at com.netflix.spinnaker.rosco.filters.SimpleCORSFilter.doFilter(SimpleCORSFilter.groovy:34) [rosco-web.jar:na]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.springframework.boot.actuate.web.trace.servlet.HttpTraceFilter.doFilterInternal(HttpTraceFilter.java:88) [spring-boot-actuator-2.1.7.RELEASE.jar:2.1.7.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:209) [spring-security-web-5.1.6.RELEASE.jar:5.1.6.RELEASE]
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178) [spring-security-web-5.1.6.RELEASE.jar:5.1.6.RELEASE]
        at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:357) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:270) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:92) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:93) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.filterAndRecordMetrics(WebMvcMetricsFilter.java:114) [spring-boot-actuator-2.1.7.RELEASE.jar:2.1.7.RELEASE]
        at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:104) [spring-boot-actuator-2.1.7.RELEASE.jar:2.1.7.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118) [spring-web-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at javax.servlet.FilterChain$doFilter.call(Unknown Source) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at com.netflix.spinnaker.filters.AuthenticatedRequestFilter.doFilter(AuthenticatedRequestFilter.groovy:140) [kork-web-7.5.1.jar:7.5.1]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:490) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:408) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:853) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1587) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_232]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_232]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-embed-core-9.0.22.jar:9.0.22]
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_232]
```
I thought that I used the wrong subPath, but I checked over an over and in the repository as in the pipeline all correct.

Then, I moved the chart to the root folder of the repository, just to see what it does, so instead of being in `src/main/resources/helm/sc-nginx/`, it was just `sc-nginx/`, I adjusted the subPath in the pipeline, and that worked fine.

So finally, I went to Rosco, and with some luck I was able to get the compressed chart that the service consumes as Clouddriver creates it, and what I saw is that instead of compressing the folder that I specified in the _subPath_ parameter, and just have `sc-nginx/`, it compresses the folder, but keeps the whole directory structure, so I end up having `src/main/resources/helm/sc-nginx`. Therefore, when you pass that to Helm, it complains that it doesn't contain a _Chart.yaml_ file (it is, but not where it expects it)

### Cloud Provider(s):
Kubernetes 1.15
Helm

### Environment:
Halyard 1.31.1 and 1.32.0
Spinnaker 1.18.4 and 1.19.1 running in Kubernetes

### Feature Area:
Clouddriver

**Comment by maggieneterval** [comment]
Hi @angelcervilla! Thanks for taking the time to open this detailed issue.

Unfortunately, the Git Repo artifact type is currently only available for Kustomize. Definitely give [this issue](https://github.com/spinnaker/spinnaker/issues/5249) a thumbs-up if you'd like to see Git Repo artifact support for Helm.

We've updated the list of artifact types available to Helm-specific Bake (Manifest) stages with [this PR](https://github.com/spinnaker/deck/pull/8053), so hopefully we can avoid this type of confusion in the future.

Thank you!

**Comment by angelcervilla** [comment]
Hi @maggieneterval 

Thank you very much for your promptly reply and for your time, appreciate it. Thanks for the link to the issue, I already gave it a thumbs-up before commenting here.

Certainly it's a pity that officially Helm charts are not supported to be consumed via Git Repo artifacts. I mean, as it is right now, it seems to work, even if this was a kind of unintended side effect, and I believe that with a few changes it'll fully work. With this, then the teams in my organization would find it very beneficial in their development cycle, as the current solutions don't fully work as:

- Get a pipeline triggered when a new chart is uploaded into Artifactory doesn't work (I raised the issue #5589 for that), and because it doesn't work, teams needs always to modify the pipeline manually and select the new chart.
- The option in the UI you created some time ago, to overwrite the Helm chart when you run the pipeline, only works when you have 1 and only 1 bake stage (our pipelines have multiple bake stages, 1 for each environment), so we cannot use it, and
- The mentioned option above seems to not works with `artifacts-rewrite` turned on, or at least I was unable to make it appear. I'll try again, if it doesn't work, then I'll raise another issue.

It's already a challenge to use Spinnaker and getting it working in our organization, and there was some dissatisfaction within the teams, so when I knew about the GitRepo artifact, and tested it (bad luck that it was with the favorable case), I believed it'll help a lot,pushed to get it upgraded, and now see it to not fully work...well, tough luck.

Again, appreciate a lot your time, and definitely it's a great idea to filter the artifact types that can be used with Helm.

Let me know if I can help you with anything else.

**Comment by angelcervilla** [comment]
Hi @maggieneterval 

Thank you very much for your promptly reply and for your time, appreciate it. Thanks for the link to the issue, I already gave it a thumbs-up before commenting here.

Certainly it's a pity that officially Helm charts are not supported to be consumed via Git Repo artifacts. I mean, as it is right now, it seems to work, even if this was a kind of unintended side effect, and I believe that with a few changes it'll fully work. With this, then the teams in my organization would find it very beneficial in their development cycle, as the current solutions don't fully work as:

- Get a pipeline triggered when a new chart is uploaded into Artifactory doesn't work (I raised the issue #5589 for that), and because it doesn't work, teams needs always to modify the pipeline manually and select the new chart.
- The option in the UI you created some time ago, to overwrite the Helm chart when you run the pipeline, only works when you have 1 and only 1 bake stage (our pipelines have multiple bake stages, 1 for each environment), so we cannot use it, and
- The mentioned option above seems to not works with `artifacts-rewrite` turned on, or at least I was unable to make it appear. I'll try again, if it doesn't work, then I'll raise another issue.

It's already a challenge to use Spinnaker and getting it working in our organization, and there was some dissatisfaction within the teams, so when I knew about the GitRepo artifact, and tested it (bad luck that it was with the favorable case), I believed it'll help a lot,pushed to get it upgraded, and now see it to not fully work...well, tough luck.

Again, appreciate a lot your time, and definitely it's a great idea to filter the artifact types that can be used with Helm.

Let me know if I can help you with anything else.
